baybe/searchspace/continuous.py

```diff
@@ -205,18 +213,86 @@ def samples_random(self, n_points: int = 1) -> pd.DataFrame:
         # without the get_polytope_samples, which means torch and botorch
         # wouldn't be needed.
 
-        points = get_polytope_samples(
-            n=n_points,
-            bounds=torch.from_numpy(self.param_bounds_comp),
-            equality_constraints=[
-                c.to_botorch(self.parameters) for c in self.constraints_lin_eq
-            ],
-            inequality_constraints=[
-                c.to_botorch(self.parameters) for c in self.constraints_lin_ineq
-            ],
-        )
+        if len(self.constraints_cardinality) != 0:
+            return self._sample_with_cardinality_constraints(n_points)
+        else:
+            points = get_polytope_samples(
+                n=n_points,
+                bounds=torch.from_numpy(self.param_bounds_comp),
+                equality_constraints=[
+                    c.to_botorch(self.parameters) for c in self.constraints_lin_eq
+                ],
+                inequality_constraints=[
+                    c.to_botorch(self.parameters) for c in self.constraints_lin_ineq
+                ],
+            )
+            return pd.DataFrame(points, columns=self.param_names)
+
+    def _sample_with_cardinality_constraints(self, n_points: int = 1) -> pd.DataFrame:
+        """Get random samples from the continuous space with cardinality constraints.
+
+        Args:
+            n_points: see sample_random()
+
+        Returns:
+            see sample_random()
+        """
+        import torch
+        from botorch.utils.sampling import get_polytope_samples
+        # TODO Revisit: torch and botorch here are actually only necessary if there
+        # are constraints. If there are none and the lists are empty we can just sample
+        # without the get_polytope_samples, which means torch and botorch
+        # wouldn't be needed. see samples_random().
+
+        if len(self.constraints_cardinality) != 0:
+            # generate all possibilities of inactive parameters which are compatible
+            # to all cardinality constraints
+            _, inactive_params_full_list = get_combinations_from_cardinality_con(
```

Comment:
Explicitly constructing this list is prohibitively expense (combinatorial blowup), and actually not needed add all, hence waste of computation. You don't need to have all combinations explicitly in memory only to then take a random selection in each iteration. Instead, just randomly select which parameters will be active! Hence, I guess your helper `get_combinations_...` becomes completely obsolete. Or is there any other reason why we need it?

Reply:
Good point! I agree that this combinatorial selection can be a computational burden. But I would say we may still need it. In the case of just one cardinality constraint, there is no need to construct such a list and we can randomly select which parameters are active. But what if we have multiple cardinality constraints? We need to generate active parameters which are valid for all cardinality constraints. If we want to randomly select which parameters to be active, we would need a "rejection sampling" strategy, ensuring us the active parameters are compatible with all cardinality constraints. Depending on the number of cardinality constraints and their complexity, this random sampling strategy can be more but also less efficient. On the other hand, if we have the list of possible candidates ready (which is computationally heavy but it is needed only once), we can randomly select one from them. Nevertheless, I believe we need to store the `inactive_params_full_list` somewhere instead of building it on each `random_sample()` call. Happy to hear your preference on 
- creating the combinatorial list (all possible (in)active parameters, which are compatible with all cardinality constraints) once and sample from them
- or randomly sample the active parameters using "rejection sampling" mechanism. Or if there is any other sampling strategy?

Reply:
discussion no longer relevant, has be fixed in the meantime